openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst

```diff
@@ -0,0 +1,102 @@
+Discussion provider configuration
+=================================
+
+
+Status
+------
+
+Proposal
+
+
+Context
+-------
+
+As part of the BD-03 initiative (Blended Development, Project 3),
+we want to enable third-party discussion providers to replace the
+default forums experience in edx-platform.
+
+To accomplish this, we should establish a configuration system to
+enable/disable/configure these new discussion-provider plugins;
+the existing/default/legacy forums experience will be included as one of these plugins.
+
+This ADR proposes a new configuration system that allows operators to create
+pre-populated configurations for specific discussions providers,
+which can then be used and customized by course authors/admins.
+
+
+Decision
+--------
+
+We propose to implement this as a Django database-backed model with
+historical records, `LearningContextDiscussionConfig`.
+
+While the underlying model will use `LearningContextKeys` (so that this
+could eventually be extended to cover (sites/programs/etc.), the first
+implementing data structure (`data.py`) can be described as:
+
+    @attr.s(frozen=True)
+    class CourseDiscussionConfigData:
+        course_key = attr.ib(type=CourseKey)
+        enabled = attr.ib(type=bool)
+        lti_configuration = attr.ib(type=dict)
+        plugin_configuration = attr.ib(type=dict)
+        provider = attr.ib(type=str)
+
+Discussions can be disabled on a course by setting `enabled=False`;
+if no record exists, the plugin is considered to be disabled by default.
+
+`provider` is the (arbitrary, but stable) key/id for the discussion provider.
```

Comment:
Nit: I'd prefer that the attributes be described in the order in which they're defined above (`enabled` then `lti_configuration` etc)

Nit: `key` makes it sound like it's supposed to be a key-value pair; I was confused for a second that you meant a key:id mapping